ass4.cpp

```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
     flood_fill_recursive(table, y, x + 1);
 }
 
-// Second variant of algorithm using data structure deque for storing islands
+int recursive_count_islands(vector<vector<bool> > table) {
+    int counter = 0;
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
+            if (table[y][x]) {
+                flood_fill_recursive(table, y, x);
+                counter++;
+            }
+        }
+    }
+    return counter;
+}
 
-void flood_fill_queued(vector<vector<bool> > &table, int y, int x) {
-    deque<pair<int, int>> tile_deque;
-    tile_deque.push_back({y,x});
+// Algorithm "Union-Find" using Tree to store disjoint components as unique islands
+
+class Disjoint_sets {
+    vector<int> rank;
+    vector<int> parent;
     
-    while (!tile_deque.empty()) {
-        pair<int, int> cur_tile = tile_deque.front();
+public:
+    Disjoint_sets(int n) {
+        rank.resize(n, 0);
+        for (int i = 0; i < n; ++i) {
+            parent.push_back(i);
+        }
+    }
     
-        tile_deque.pop_front();
+    int find(int x) {
+        if (parent[x] != x) {
+            return find(parent[x]);
```

Comment:
It's a very common optimization to set `parent[x] = find(parent[x])` this way the tree gets shallower and it's faster the next time around.